### PR TITLE
refactor dataflow code and print range info

### PR DESCRIPTION
**Pull request by jubitaneja**

### CODE REVIEW

**Comment by jubitaneja** [comment]
This patch refactors the dataflow code, and also prints range information as computed by LLVM in the extractor. This was missing from the previous PR on constant range.

**Comment by jubitaneja** [comment]
@zhengyangl Please review this as well. It refactors the code and adds the missing case for printing range computed by LLVM.

File: lib/Extractor/Candidates.cpp [hunk 1]
```diff
@@ -860,6 +864,70 @@ std::tuple<BlockPCs, std::vector<InstMapping>> GetRelevantPCs(
...
 }
 
+void PrintDataflowInfo(Function &F, Instruction &I, LazyValueInfo *LVI,
+                       ScalarEvolution *SE) {
+  auto V = I.getOperand(0);
+  auto DL = F.getParent()->getDataLayout();
+  if (PrintNegAtReturn) {
+    bool Negative = isKnownNegative(V, DL);
+    if (Negative)
+      llvm::outs() << "known negative from compiler: " << "true" << "\n";
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: don't keep duplicating strings like this, just print the first part of this before the conditional

File: lib/Extractor/Candidates.cpp [hunk 2]
```diff
@@ -860,6 +864,70 @@ std::tuple<BlockPCs, std::vector<InstMapping>> GetRelevantPCs(
...
+      if (Instruction *I = dyn_cast<Instruction>(V)) {
+        BasicBlock *BB = I->getParent();
+        auto LVIRange = LVI->getConstantRange(V, BB);
+        auto SC = SE->getSCEV(V);
+        auto R1 = LVIRange.intersectWith(SE->getSignedRange(SC));
+        auto R2 = LVIRange.intersectWith(SE->getUnsignedRange(SC));
+        Range = R1.getSetSize().ult(R2.getSetSize()) ? R1 : R2;
+      }
+    }
+    llvm::outs() << "known range from compiler: [" << Range.getLower() << ","
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: just range, not "known range"

File: lib/Extractor/Candidates.cpp [hunk 3]
```diff
@@ -860,6 +864,70 @@ std::tuple<BlockPCs, std::vector<InstMapping>> GetRelevantPCs(
...
+    bool NonZero = isKnownNonZero(V, DL);
+    if (NonZero)
+      llvm::outs() << "known nonZero from compiler: " << "true" << "\n";
+    else
+      llvm::outs() << "known nonZero from compiler: " << "false" << "\n";
+  }
+  if (PrintSignBitsAtReturn) {
+    unsigned NumSignBits = ComputeNumSignBits(V, DL);
+    if (NumSignBits > 1)
+      llvm::outs() << "known signBits from compiler: " << NumSignBits << "\n";
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: only say "known" for the actual knownBits analysis, it's part of the name of that analysis, but it's not part of the name of the other ones

Comment (jubitaneja) [reply lib/Extractor/Candidates.cpp]: sure, will fix it. I used known as a verb here, but I agree it gets confusing with the actual analysis name.

File: lib/Extractor/Candidates.cpp [hunk 4]
```diff
@@ -861,6 +865,75 @@ std::tuple<BlockPCs, std::vector<InstMapping>> GetRelevantPCs(
   return std::make_tuple(RelevantBPCs, RelevantPCs);
 }
 
+void PrintDataflowInfo(Function &F, Instruction &I, LazyValueInfo *LVI,
+                       ScalarEvolution *SE) {
+  auto V = I.getOperand(0);
+  auto DL = F.getParent()->getDataLayout();
+  if (PrintNegAtReturn) {
+    bool Negative = isKnownNegative(V, DL);
+    std::string s = "negative from compiler: ";
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: fine but why not just print this right here? why make another variable??

File: lib/Extractor/Candidates.cpp [hunk 5]
```diff
@@ -861,6 +865,75 @@ std::tuple<BlockPCs, std::vector<InstMapping>> GetRelevantPCs(
...
+    std::string s = "nonZero from compiler: ";
+    if (NonZero)
+      llvm::outs() << s << "true" << "\n";
+    else
+      llvm::outs() << s << "false" << "\n";
+  }
+  if (PrintSignBitsAtReturn) {
+    unsigned NumSignBits = ComputeNumSignBits(V, DL);
+    std::string s = "signBits from compiler: ";
+    if (NumSignBits > 1)
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: why is this test needed?

File: lib/Extractor/Candidates.cpp [hunk 6]
```diff
@@ -861,6 +865,71 @@ std::tuple<BlockPCs, std::vector<InstMapping>> GetRelevantPCs(
...
 
+void PrintDataflowInfo(Function &F, Instruction &I, LazyValueInfo *LVI,
+                       ScalarEvolution *SE) {
+  auto V = I.getOperand(0);
+  auto DL = F.getParent()->getDataLayout();
+  if (PrintNegAtReturn) {
+    bool Negative = isKnownNegative(V, DL);
+    llvm::outs() << "negative from compiler: ";
+    if (Negative)
+      llvm::outs() << "true\n";
```
Comment (regehr) [inline lib/Extractor/Candidates.cpp]: ok this is getting there, thanks.

there's one more very easy way to simplify this code, please see here: http://www.cplusplus.com/reference/ios/boolalpha/

just set this flag at the start of PrintDataflowInfo() and clear it at the end

Comment (jubitaneja) [reply lib/Extractor/Candidates.cpp]: Thanks for sharing this flag. Hope I have used it the right way.

Comment (zhengyangl) [reply lib/Extractor/Candidates.cpp]: @regehr std::boolalpha won't work together with llvm::errs() and outs().

**Comment by regehr** [comment]
ok, good
see how much cleaner and nicer the code is when you do this right?

**Comment by jubitaneja** [comment]
Yes. :)

**Pull request merged**